File: JavaScript/0937. Reorder Data in Log Files.js
// You have an array of logs.  Each log is a space delimited string of words.
// For each log, the first word in each log is an alphanumeric identifier.  Then, either:
//
// - Each word after the identifier will consist only of lowercase letters, or;
// - Each word after the identifier will consist only of digits.
//
// We will call these two varieties of logs letter-logs and digit-logs. It is guaranteed that each log has at least one word after its identifier.
// Reorder the logs so that all of the letter-logs come before any digit-log. The letter-logs are ordered lexicographically ignoring identifier, with the identifier used in case of ties.  The digit-logs should be put in their original order.
// Return the final order of the logs.
//
// Example 1:
//
// Input: logs = ["dig1 8 1 5 1","let1 art can","dig2 3 6","let2 own kit dig","let3 art zero"]
// Output: ["let1 art can","let3 art zero","let2 own kit dig","dig1 8 1 5 1","dig2 3 6"]
//
// Constraints:
//
// 0 <= logs.length <= 100
// 3 <= logs[i].length <= 100
// logs[i] is guaranteed to have an identifier, and a word after the identifier.

/**
 * @param {string[]} logs
 * @return {string[]}
 */
function reorderLogFiles(logs) {
  const body = s => s.slice(s.indexOf(' ') + 1); // get the body after identifier
  const isNumber = c => /\d/.test(c);

  // if body same then compare identifier
  const compare = (a, b) => {
    const n = body(a).localeCompare(body(b));
    if (n !== 0) return n;
    return a.localeCompare(b);
  };

  const digitLogs = [];
  const letterLogs = [];
  for (const log of logs) {
    if (isNumber(body(log))) {
      digitLogs.push(log);
    } else {
      letterLogs.push(log);
    }
  }
  return [...letterLogs.sort(compare), ...digitLogs];
}

File: Python/0937. Reorder Data in Log Files.py
# You are given an array of logs. Each log is a space-delimited string of words, where the first word is the identifier.
# There are two types of logs:
# - Letter-logs: All words (except the identifier) consist of lowercase English letters.
# - Digit-logs: All words (except the identifier) consist of digits.
# Reorder these logs so that:
# 1. The letter-logs come before all digit-logs.
# 2. The letter-logs are sorted lexicographically by their contents. If their contents are the same, then sort them lexicographically by their identifiers.
# 3. The digit-logs maintain their relative ordering.
# Return the final order of the logs.
#
# Example 1:
#
# Input: logs = ["dig1 8 1 5 1","let1 art can","dig2 3 6","let2 own kit dig","let3 art zero"]
# Output: ["let1 art can","let3 art zero","let2 own kit dig","dig1 8 1 5 1","dig2 3 6"]
# Explanation:
# The letter-log contents are all different, so their ordering is "art can", "art zero", "own kit dig".
# The digit-logs have a relative order of "dig1 8 1 5 1", "dig2 3 6".
#
# Example 2:
#
# Input: logs = ["a1 9 2 3 1","g1 act car","zo4 4 7","ab1 off key dog","a8 act zoo"]
# Output: ["g1 act car","a8 act zoo","ab1 off key dog","a1 9 2 3 1","zo4 4 7"]
#
# Constraints:
#
# 1 <= logs.length <= 100
# 3 <= logs[i].length <= 100
# All the tokens of logs[i] are separated by a single space.
# logs[i] is guaranteed to have an identifier and at least one word after the identifier.


class Solution:
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        return sorted(logs, key=self.compare)

    def compare(self, logs):
        identifier, s = logs.split(" ", 1)  # 1 is max split times
        if s[0].isalpha():
            return (0, s, identifier)
        else:
            return (1, None, None)
